### PR TITLE
Task 3

### PR DESCRIPTION
**Pull request by Vladlan**
**1. What was done?**
 
- **+1** - serverless.ts contains configuration for 2 lambda functions, API is not working at all, but configuration is correct
- **+1** - The getAllProducts OR getProductsById lambda function returns a correct response (POINT1)
- **+1** - The getProductsById AND getProductsList lambda functions return a correct response code (POINT2)
- **+1** - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. 

**Additional (optional) tasks (but nice to have):**
- **+1** - Async/await is used in lambda functions
- **+1** - ES6 modules are used for product-service implementation
- **+1** - Webpack is configured for product-service
- **+1** - SWAGGER documentation is created for product-service
- **+1** - Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) (You may use JEST)
- **+1** - Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase.
- **+1** - Main error scenarios are handled by API ("Product not found" error, try catch blocks are used in lambda handlers).
   
2. Link to product-service API - https://269pjntj1e.execute-api.eu-west-1.amazonaws.com/dev/products
3. Link to FE MR (YOUR OWN REPOSITORY) - https://github.com/Vladlan/nodejs-aws-fe/pull/2 (https://d32m2krhvblmxn.cloudfront.net/)
4. Link to SWAGGER - https://app.swaggerhub.com/apis/nodejs-aws-be/nodejs-aws-be/1.0.0#/info

**Total score: 11**

### CODE REVIEW

File: products-service/handlers/get-product-by-id.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                };
+            }
+            return {
+                statusCode: 404,
+                headers: corsHeaders,
+                body: errorMessages.notFound(productId),
+            }
+        }
+        return {
+            statusCode: 400,
```
Comment (Emerkof) [inline products-service/handlers/get-product-by-id.ts:25]: This error will never be triggered. If you send a request to your `products/{id}` route without an `{id}` param, you'll simply trigger a `/products` route handler, which will return you a list of products.

File: products-service/handlers/get-product-by-id.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { APIGatewayProxyHandler } from 'aws-lambda';
+import * as products from '../products/products.json';
+import 'source-map-support/register';
+import { corsHeaders, errorMessages } from '../utils';
+
+export const getProductById: APIGatewayProxyHandler = async (event, _context) => {
+    try {
```
Comment (Emerkof) [inline products-service/handlers/get-product-by-id.ts:7]: I'd suggest moving duplicate logic (like error handling, response formatting) into separate lib/ functions and use them across the handlers.

It's a wish for the future, no need to implement it right now. It won't affect your score.

File: products-service/handlers/get-product-by-id.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        if (productId) {
+            const product = products.find(({ id }) => productId === id);
+            if (product) {
+                return {
+                    statusCode: 200,
+                    headers: corsHeaders,
+                    body: JSON.stringify(product, null, 2),
+                };
+            }
+            return {
```
Comment (Emerkof) [inline products-service/handlers/get-product-by-id.ts:18]: It'd be better to throw an error here (for example, custom `NotFound` error) and handle all the errors in 1 place (error handler function).

It's a wish for the future, no need to implement it right now. It won't affect your score.

File: products-service/handlers/get-product-by-id.ts (hunk 1 above)
Comment (Vladlan) [reply products-service/handlers/get-product-by-id.ts:25]: Agree, but no one can guarantee that endpoint for getProductById will not change, but I can guarantee that after that change we will forget to add error handling.

File: products-service/handlers/get-product-by-id.ts (hunk 2 above)
Comment (Vladlan) [reply products-service/handlers/get-product-by-id.ts:7]: Agreed

File: products-service/handlers/get-product-by-id.ts (hunk 3 above)
Comment (Vladlan) [reply products-service/handlers/get-product-by-id.ts:18]: Agreed

**Review by ViktarYarmalovich — APPROVED** [review]

**Pull request merged**